storagenode/piecestore/retain.go

```diff
@@ -0,0 +1,186 @@
+		retainStatus: retainStatus,
+		queued:       make(map[storj.NodeID]RetainRequest),
+		reqChan:      make(chan RetainRequest),
+		sem:          make(chan struct{}, concurrentRetain),
+		emptyTrigger: make(chan struct{}),
+		store:        store,
+	}
+}
+
+// QueueRetain adds a retain request to the queue
+func (s *RetainService) QueueRetain(req RetainRequest) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// only queue retain request if we do not already have one for this satellite
+	if _, ok := s.queued[req.SatelliteID]; !ok {
+		s.queued[req.SatelliteID] = req
+		go func() { s.reqChan <- req }()
+	}
+}
+
+// Run listens for queued retain requests and processes them as they come in
+func (s *RetainService) Run(ctx context.Context) error {
+	for {
+		// exit if context has been canceled. Otherwise, block until an item can be added to the semaphore
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case s.sem <- struct{}{}:
+		}
+
+		// get the next request
+		var req RetainRequest
+		select {
+		case req = <-s.reqChan:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		// TODO make it possible to sync with this goroutine
+		go func(ctx context.Context, req RetainRequest) {
+			err := s.retainPieces(ctx, req)
+			if err != nil {
+				s.log.Error("retain error", zap.Error(err))
+			}
+			s.mu.Lock()
+			delete(s.queued, req.SatelliteID)
+			s.mu.Unlock()
+
+			if len(s.queued) == 0 {
+				s.emptyTrigger <- struct{}{}
+			}
+
+			// remove item from semaphore and free up process for another retain job
+			<-s.sem
+		}(ctx, req)
+	}
+}
+
+// Wait blocks until the context is canceled or until the queue is empty
+func (s *RetainService) Wait(ctx context.Context) {
+	s.mu.Lock()
+	queueLength := len(s.queued)
+	s.mu.Unlock()
+	if queueLength == 0 {
+		return
+	}
+	select {
+	case <-s.emptyTrigger:
+	case <-ctx.Done():
+	}
+}
+
+func (s *RetainService) retainPieces(ctx context.Context, req RetainRequest) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// if retain status is disabled, return immediately
+	if s.retainStatus == RetainDisabled {
+		return nil
+	}
+
+	numDeleted := 0
+	satelliteID := req.SatelliteID
+	filter := req.Filter
+	createdBefore := req.CreatedBefore
+
+	s.log.Info("Prepared to run a Retain request.",
+		zap.Time("createdBefore", createdBefore),
+		zap.Int64("filterSize", filter.Size()),
+		zap.String("satellite", satelliteID.String()))
+
+	err = s.store.WalkSatellitePieces(ctx, satelliteID, func(access pieces.StoredPieceAccess) error {
+		// We call Gosched() when done because the GC process is expected to be long and we want to keep it at low priority,
+		// so other goroutines can continue serving requests.
+		defer runtime.Gosched()
+		// See the comment above the Retain() function for a discussion on the correctness
+		// of using ModTime in place of the more precise CreationTime.
+		mTime, err := access.ModTime(ctx)
+		if err != nil {
+			s.log.Error("failed to determine mtime of blob", zap.Error(err))
```

Comment:
I would change this level to a `WARN`.
I've seen that SNO gets alarmed on any ERROR message, hence it would avoid having them to worry a lot if this happens, considering that the process carries on with the next one.

Reply:
Fixed